File: literallytoast/src/main/java/com/dvoiss/literallytoast/Toast.java
package com.dvoiss.literallytoast;

import android.annotation.SuppressLint;
import android.content.Context;
import android.graphics.Typeface;
import android.os.Handler;
import android.support.annotation.IntDef;
import android.support.annotation.NonNull;
import android.support.annotation.Nullable;
import android.support.annotation.StringRes;
import android.view.LayoutInflater;
import android.view.View;
import android.view.WindowManager;
import android.widget.TextView;
import com.facebook.rebound.SpringSystem;
import java.lang.annotation.Retention;
import java.lang.annotation.RetentionPolicy;

public class Toast {
    @Nullable private static Typeface typeface;
    @Nullable private static WindowManager windowManager;

    @NonNull private final View view;
    @Duration private final int duration;

    @NonNull private final Handler handler;

    @SuppressLint("InflateParams")
    private Toast(@NonNull Context context, @NonNull CharSequence text, @Duration int duration) {
        this.duration = duration;
        this.view = getLayoutInflater(context).inflate(R.layout.literallytoast, null);
        this.handler = new Handler();
        setupStyledTextView(context, this.view, text);
    }

    @NonNull
    public static Toast create(@NonNull Context context, @NonNull CharSequence text,
        @Duration int duration) {
        return new Toast(context, text, duration);
    }

    @NonNull
    public static Toast create(@NonNull Context context, @StringRes int resId,
        @Duration int duration) {
        return create(context, context.getString(resId), duration);
    }

    public void show() {
        getWindowManager(view.getContext()).addView(view, ToastUtils.createWindowManagerParams());
        SpringSystem.create()
            .createSpring()
            .addListener(new ToastSpringListener(view, windowManager, handler, getDuration()))
            .setEndValue(1);
    }

    private long getDuration() {
        if (duration == LENGTH_SHORT) {
            return SHORT_DURATION_TIMEOUT;
        } else if (duration == LENGTH_LONG) {
            return LONG_DURATION_TIMEOUT;
        } else {
            return duration;
        }
    }

    private static void setupStyledTextView(@NonNull Context context, @NonNull View view,
        @NonNull CharSequence text) {
        TextView textView = view.findViewById(R.id.message);
        textView.setTypeface(getTypeface(context));
        textView.setText(text);
    }

    @NonNull
    private static LayoutInflater getLayoutInflater(@NonNull Context context) {
        return (LayoutInflater) context.getSystemService(Context.LAYOUT_INFLATER_SERVICE);
    }

    @NonNull
    private static Typeface getTypeface(@NonNull Context context) {
        if (typeface == null) {
            typeface = Typeface.createFromAsset(context.getAssets(), "fonts/yummy_bread.ttf");
        }
        return typeface;
    }

    @NonNull
    private static WindowManager getWindowManager(@NonNull Context context) {
        if (windowManager == null) {
            windowManager = (WindowManager) context.getApplicationContext()
                .getSystemService(Context.WINDOW_SERVICE);

            if (windowManager == null) {
                throw new IllegalStateException("Could not find window manager.");
            }
        }

        return windowManager;
    }

    // region Duration IntDef

    /**
     * Duration copied from android.widget.Toast.
     */
    @IntDef({ LENGTH_SHORT, LENGTH_LONG })
    @Retention(RetentionPolicy.SOURCE)
    public @interface Duration {}

    /**
     * Copied from android.widget.Toast: {@link android.widget.Toast#LENGTH_SHORT}.
     */
    public static final int LENGTH_SHORT = 0;

    /**
     * Copied from android.widget.Toast: {@link android.widget.Toast#LENGTH_LONG}.
     */
    public static final int LENGTH_LONG = 1;

    /**
     * Copied from android.widget.Toast.
     */
    static final long SHORT_DURATION_TIMEOUT = 4000;

    /**
     * Copied from android.widget.Toast.
     */
    static final long LONG_DURATION_TIMEOUT = 7000;

    // endregion
}

File: literallytoast/src/main/java/com/dvoiss/literallytoast/Toast.kt
package com.dvoiss.literallytoast

import android.content.Context
import android.graphics.Typeface
import android.media.AudioManager
import android.media.MediaPlayer
import android.os.Handler
import android.support.annotation.StringRes
import android.view.LayoutInflater
import android.view.View
import android.view.WindowManager
import android.widget.TextView
import com.facebook.rebound.SpringSystem


class Toast private constructor(context: Context, text: CharSequence, private val duration: Int,
    private var shouldPlaySound: Boolean) {

  private val view: View
  private val handler: Handler

  init {
    this.view = getLayoutInflater(context).inflate(R.layout.literallytoast, null)
    this.handler = Handler()
    setupStyledTextView(context, this.view, text)
  }

  fun show() {
    

    getWindowManager(view.context).addView(view, ToastUtils.createWindowManagerParams())
    SpringSystem.create()
        .createSpring()
        .addListener(
            ToastSpringListener(view, windowManager!!, handler, getDuration())).endValue = 1.0

    if (shouldPlaySound) {
      val audioManager = view.context.getSystemService(Context.AUDIO_SERVICE) as AudioManager
      val originalVolume = audioManager.getStreamVolume(AudioManager.STREAM_MUSIC)
      audioManager.setStreamVolume(AudioManager.STREAM_MUSIC, 20, 0)
      val mp = MediaPlayer.create(view.context, R.raw.toaster)
      mp.setOnCompletionListener { it.release() }
      mp.start()
      audioManager.setStreamVolume(AudioManager.STREAM_MUSIC, originalVolume, 0)
    }
  }

  fun setPlaySound(shouldPlaySound: Boolean): Toast {
    this.shouldPlaySound = shouldPlaySound
    return this
  }

  private fun getDuration(): Long {
    return if (duration == LENGTH_SHORT) {
      SHORT_DURATION_TIMEOUT
    } else if (duration == LENGTH_LONG) {
      LONG_DURATION_TIMEOUT
    } else {
      duration.toLong()
    }
  }

  private fun setupStyledTextView(context: Context, view: View, text: CharSequence) {
    val textView = view.findViewById<TextView>(R.id.message)
    textView.typeface = getTypeface(context)
    textView.text = text
  }

  private fun getLayoutInflater(context: Context): LayoutInflater = context.getSystemService(
      Context.LAYOUT_INFLATER_SERVICE) as LayoutInflater

  companion object {
    private var typeface: Typeface? = null
    private var windowManager: WindowManager? = null

    @JvmStatic
    @JvmOverloads
    fun create(context: Context, text: CharSequence, duration: Int = LENGTH_SHORT): Toast =
        Toast(context, text, duration, shouldPlaySound = true)

    @JvmStatic
    @JvmOverloads
    fun create(context: Context, @StringRes resId: Int,
        duration: Int = LENGTH_SHORT): Toast = create(context, context.getString(resId), duration)

    private fun getTypeface(context: Context): Typeface = typeface ?: Typeface.createFromAsset(
        context.assets, "fonts/yummy_bread.ttf")

    private fun getWindowManager(context: Context): WindowManager {
      if (windowManager == null) {
        windowManager = context.applicationContext
            .getSystemService(Context.WINDOW_SERVICE) as WindowManager

        if (windowManager == null) {
          throw IllegalStateException("Could not find window manager.")
        }
      }

      return windowManager as WindowManager
    }

    // region The values in this region were copied from android.widget.Toast.

    const val LENGTH_SHORT = 0
    const val LENGTH_LONG = 1
    internal val SHORT_DURATION_TIMEOUT: Long = 4000
    internal val LONG_DURATION_TIMEOUT: Long = 7000

    // endregion
  }
}
